src/legacy/reduce.rs

```diff
@@ -0,0 +1,314 @@
+use std::collections::VecDeque;
+use std::io::{self, copy, Read, Result};
+
+use crate::legacy::lz77::lz77_output_backref;
+
+use super::bitstream::{lsb, BitStream};
+
+/// Number of bits used to represent indices in a follower set of size n.
+fn follower_idx_bw(n: u8) -> u8 {
+    debug_assert!(n <= 32);
+    match n {
+        0 => 0,
+        1 => 1,
+        _ => {
+            5 - ((n - 1) << 3).leading_zeros() as u8
+        }
+    }
+}
+
+#[derive(Default, Clone, Copy)]
+struct FollowerSet {
+    size: u8,
+    idx_bw: u8,
+    followers: [u8; 32],
+}
+
+/// Read the follower sets from is into fsets. Returns true on success.
+fn read_follower_sets(is: &mut BitStream, fsets: &mut [FollowerSet]) -> io::Result<()> {
+    for i in (0..=u8::MAX as usize).rev() {
+        let n = lsb(is.bits(), 6) as u8;
```

Comment:
Combine this and the `advance()` call into a `read_next_bits` function, so nobody can forget to include the `advance()`. In fact, do this wherever `lsb` and `advance` are applied to the same number of bits.

Reply:
Makes sense most of the time - there is just one case where it's a bit more complex to advance.